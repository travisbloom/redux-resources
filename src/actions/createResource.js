import { capitalize, generateConstantFromString } from '../utilities'
import {
    CREATE_RESOURCE_REQUEST,
    CREATE_RESOURCE_REQUEST_SUCCESS,
    CREATE_RESOURCE_REQUEST_ERROR,
} from '../constants'

const createResource = ({ formatErrors, request, resource, normalizer }) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `CREATE_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const actionCreatorName = `create${capitalize(resource)}`

    const createResourceActionCreator = (body, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors } = options
        const meta = { ...options, timestamp: new Date().toISOString(), resource }
        dispatch({
            type: initialActionType,
            payload: body,
            meta: { ...meta, reduxResourcesActionType: CREATE_RESOURCE_REQUEST },
        })
        return request(body, { ...options, dispatch, getState }).then((response) => {
            dispatch({
                type: successActionType,
                payload: response,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(response),
                    reduxResourcesActionType: CREATE_RESOURCE_REQUEST_SUCCESS,
                },
            })
            return response
        })
        .catch((response) => {
            dispatch({
                type: errorActionType,
                payload: formatErrors(response),
                meta: { ...meta, reduxResourcesActionType: CREATE_RESOURCE_REQUEST_ERROR },
            })
            if (shouldThrowErrors) throw response
        })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [actionCreatorName]: createResourceActionCreator,
    }
}

export default createResource
