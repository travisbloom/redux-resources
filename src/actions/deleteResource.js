import { capitalize, generateConstantFromString } from '../utilities'
import {
    DELETE_RESOURCE_REQUEST,
    DELETE_RESOURCE_REQUEST_SUCCESS,
    DELETE_RESOURCE_REQUEST_ERROR,
} from '../constants'

const deleteResource = ({ formatErrors, request, resource }) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `DELETE_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const actionCreatorName = `delete${capitalize(resource)}`

    const deleteResourceActionCreator = (id, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors } = options
        const meta = { ...options, resource, timestamp: new Date().toISOString(), id }
        dispatch({
            type: initialActionType,
            payload: id,
            meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST },
        })
        return request(id, { ...options, dispatch, getState })
            .then(() => {
                dispatch({
                    type: successActionType,
                    payload: id,
                    meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST_SUCCESS },
                })
            })
            .catch((response) => {
                dispatch({
                    type: errorActionType,
                    payload: formatErrors(response),
                    meta: { ...meta, reduxResourcesActionType: DELETE_RESOURCE_REQUEST_ERROR },
                })
                if (shouldThrowErrors) throw response
            })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [actionCreatorName]: deleteResourceActionCreator,
    }
}

export default deleteResource
