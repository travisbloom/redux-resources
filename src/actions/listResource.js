import { serializeOrderedQuery, capitalize, generateConstantFromString } from '../utilities'
import {
    selectResourceList,
    selectSelectedResourceList,
    selectResourceListsBeingFetched,
} from '../selectors'
import {
    LIST_RESOURCE_REQUEST,
    LIST_RESOURCE_REQUEST_SUCCESS,
    LIST_RESOURCE_REQUEST_ERROR,
    SELECT_RESOURCE_LIST,
    DEFAULT_QUERY_LIMIT,
} from '../constants'

const defaultIsCached = obj => !!obj

const listResource = ({
    formatErrors,
    resource,
    request,
    normalizer,
    isCachedFn,
    stringifyQueryFn,
}) => {
    const resourceActionTypeName = generateConstantFromString(resource)
    const initialActionType = `LIST_${resourceActionTypeName}_REQUEST`
    const successActionType = `${initialActionType}_SUCCESS`
    const errorActionType = `${initialActionType}_ERROR`
    const selectActionType = `SELECT_${resourceActionTypeName}_LIST`
    const actionCreatorName = `list${capitalize(resource)}`

    const isCached = isCachedFn || defaultIsCached
    const stringifyQuery = stringifyQueryFn || serializeOrderedQuery
    /**
    * @param {object} the query param object being used to filter the resource request
    * @param {object} the options object
    *     @property {boolean} shouldIgnoreCache - if true, will fetch the remote resource even if it
          has been previously retrieved
    *     @property {boolean} shouldThrowErrors - if true, will not pass returned errors to redux
          but instead throw them so the fn calling can .catch them
    */
    const listResourceActionCreator = (passedQuery = {}, options = {}) => (dispatch, getState) => {
        const { shouldThrowErrors, shouldIgnoreCache } = options
        const query = {
            limit: DEFAULT_QUERY_LIMIT,
            offset: 0,
            ...passedQuery,
        }

        const queryString = stringifyQuery(query)
        const state = getState()
        const previousResourceList = selectResourceList(state, resource, queryString)
        const meta = {
            query,
            queryString,
            resource,
            timestamp: new Date().toISOString(),
            ...options,
        }

        // if the resource already exists in redux
        if (!shouldIgnoreCache && isCached(previousResourceList, query, options)) {
            // if the requested resource is not the current selected resource
            if (selectSelectedResourceList(state, resource) !== queryString) {
                dispatch({
                    type: selectActionType,
                    payload: queryString,
                    meta: { ...meta, reduxResourcesActionType: SELECT_RESOURCE_LIST },
                })
            }
            return Promise.resolve()
        }

        // if the resource is currently being fetched already
        if (selectResourceListsBeingFetched(state, resource).includes(queryString)) {
            return Promise.resolve()
        }

        dispatch({
            type: initialActionType,
            payload: queryString,
            meta: { ...meta, reduxResourcesActionType: LIST_RESOURCE_REQUEST },
        })

        return request(query, { ...options, dispatch, getState }).then((response) => {
            dispatch({
                type: successActionType,
                payload: response,
                meta: {
                    ...meta,
                    normalizedResponse: normalizer(response),
                    reduxResourcesActionType: LIST_RESOURCE_REQUEST_SUCCESS,
                },
            })
        })
        .catch((response) => {
            dispatch({
                type: errorActionType,
                payload: formatErrors(response),
                meta: {
                    ...meta,
                    reduxResourcesActionType: LIST_RESOURCE_REQUEST_ERROR,
                },
            })
            if (shouldThrowErrors) throw response
        })
    }
    return {
        [initialActionType]: initialActionType,
        [successActionType]: successActionType,
        [errorActionType]: errorActionType,
        [selectActionType]: selectActionType,
        [actionCreatorName]: listResourceActionCreator,
    }
}

export default listResource
