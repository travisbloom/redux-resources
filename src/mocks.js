import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Schema, normalize, arrayOf } from 'normalizr'

const RESOURCE_REDUCER = 'resources'

const RESOURCE_ID = 'resourceId'
const RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
}

const RESOURCE_UPDATE_PAYLOAD = {
    id: RESOURCE_ID,
    prop1: 'updated prop1',
    prop2: 'updated prop1',
}

const SERVER_ERRORS = [
    'This is an initial error',
    'This is another error',
]

const resourceSchema = new Schema(RESOURCE_REDUCER)
const normalizer = (resources) => {
    const array = Array.isArray(resources) ? resources : [resources]
    return normalize(array, arrayOf(resourceSchema))
}
const formatErrors = errors => errors

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

module.exports = {
    RESOURCE_REDUCER,
    RESOURCE_ID,
    RESOURCE,

    RESOURCE_UPDATE_PAYLOAD,

    SERVER_ERRORS,

    mockStore,
    formatErrors,
    normalizer,
}
