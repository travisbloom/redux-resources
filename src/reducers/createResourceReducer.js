import {
    CREATE_RESOURCE_REQUEST,
    CREATE_RESOURCE_REQUEST_ERROR,
    CREATE_RESOURCE_REQUEST_SUCCESS,
} from '../constants'

const createResourceReducer = (state, action) => {
    const { payload, meta } = action
    const { resource, timestamp, reduxResourcesActionType, normalizedResponse } = meta
    const resourceState = state[resource]

    switch (reduxResourcesActionType) {
    case CREATE_RESOURCE_REQUEST:
        return {
            ...state,
            [resource]: {
                ...resourceState,
                isFetching: true,
            },
        }
    case CREATE_RESOURCE_REQUEST_SUCCESS: {
        const { entities, result } = normalizedResponse
        return Object.keys(state).reduce((newState, resourceKey) => {
            if (resourceKey !== resource && entities[resourceKey]) {
                newState[resourceKey] = {
                    ...state[resourceKey],
                    resources: {
                        ...state[resourceKey].resources,
                        ...entities[resourceKey],
                    },
                }
            } else if (resourceKey === resource) {
                newState[resourceKey] = {
                    ...state[resourceKey],
                    isFetching: false,
                    errors: [],
                    selectedResource: result[0],
                    lastCreatedAt: timestamp,
                    resources: {
                        ...state[resourceKey].resources,
                        ...entities[resourceKey],
                    },
                }
            } else {
                newState[resourceKey] = state[resourceKey]
            }
            return newState
        }, {})
    }
    case CREATE_RESOURCE_REQUEST_ERROR:
        return {
            ...state,
            isFetching: false,
            [resource]: {
                ...resourceState,
                isFetching: false,
                errors: [
                    ...resourceState.errors,
                    ...payload,
                ],
            },
        }
    }
}

export default createResourceReducer
