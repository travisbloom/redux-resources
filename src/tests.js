import MockDate from 'mockdate'

MockDate.set(new Date('2016-01-01T00:00:00.000Z'))
