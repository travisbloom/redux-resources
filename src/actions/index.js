import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import retrieveResource from './retrieveResource'
import updateResource from './updateResource'

module.exports = {
    createResource,
    listResource,
    retrieveResource,
    updateResource,
    deleteResource,
}
