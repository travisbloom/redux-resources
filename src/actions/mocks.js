import {
    mockStore,
    normalizer,
    RESOURCE_REDUCER,
    RESOURCE,
    RESOURCE_ID,
    INITIAL_STATE,
    INITIAL_STATE_WITH_CACHED_RESOURCE,
    QUERY,
    RESOURCE_UPDATE_PAYLOAD,
    formatErrors,
} from '../mocks'
import createResource from './createResource'
import deleteResource from './deleteResource'
import listResource from './listResource'
import partialUpdateResource from './partialUpdateResource'
import updateResource from './updateResource'
import retrieveResource from './retrieveResource'

const generateCreateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState)
    const { createMockResources } = createResource({
        resource: RESOURCE_REDUCER,
        normalizer,
        request,
        formatErrors,
    })
    await store.dispatch(createMockResources(RESOURCE, options))
    return store.getActions()
}

const generateDeleteResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { deleteMockResources } = deleteResource({
        resource: RESOURCE_REDUCER,
        request,
        formatErrors,
    })
    await store.dispatch(deleteMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generateListResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { listMockResources } = listResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(listMockResources(QUERY, options))
    return store.getActions()
}

const generateRetrieveResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE)
    const { retrieveMockResources } = retrieveResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(retrieveMockResources(RESOURCE_ID, options))
    return store.getActions()
}

const generatePartialUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { partialUpdateMockResources } = partialUpdateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    await store.dispatch(partialUpdateMockResources(RESOURCE_ID, RESOURCE_UPDATE_PAYLOAD, options))
    return store.getActions()
}

const generateUpdateResourceActions = async ({ request, initialState, ...options }) => {
    const store = mockStore(initialState || INITIAL_STATE_WITH_CACHED_RESOURCE)
    const { updateMockResources } = updateResource({
        resource: RESOURCE_REDUCER,
        request,
        normalizer,
        formatErrors,
    })
    const body = { ...RESOURCE, ...RESOURCE_UPDATE_PAYLOAD }
    await store.dispatch(updateMockResources(RESOURCE_ID, body, options))
    return store.getActions()
}

module.exports = {
    generateCreateResourceActions,
    generateDeleteResourceActions,
    generateListResourceActions,
    generateRetrieveResourceActions,
    generatePartialUpdateResourceActions,
    generateUpdateResourceActions,
}
