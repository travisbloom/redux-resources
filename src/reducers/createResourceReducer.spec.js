import { RESOURCE, SERVER_ERRORS } from '../mocks'
import { generateCreateResourceActions } from '../actions/mocks'
import { snap } from '../tests'

import { getResourceInitialState } from './'
import createResourceReducer from './createResourceReducer'

it('will update the state after the initial action', async () => {
    const [initialAction] = await generateCreateResourceActions({
        request: () => Promise.resolve(RESOURCE),
    })
    snap(createResourceReducer(getResourceInitialState(), initialAction))
})

it('will update the state after the success action', async () => {
    const [initialAction, successAction] = await generateCreateResourceActions({
        request: () => Promise.resolve(RESOURCE),
    })
    const stateAfterInititalAction = createResourceReducer(getResourceInitialState(), initialAction)
    snap(createResourceReducer(stateAfterInititalAction, successAction))
})

it('will update the state after the error action', async () => {
    const [initialAction, errorAction] = await generateCreateResourceActions({
        request: () => Promise.reject(SERVER_ERRORS),
    })
    const stateAfterInititalAction = createResourceReducer(getResourceInitialState(), initialAction)
    snap(createResourceReducer(stateAfterInititalAction, errorAction))
})
