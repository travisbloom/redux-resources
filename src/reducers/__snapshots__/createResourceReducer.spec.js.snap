exports[`test will update the state after the error action 1`] = `
Object {
  "isFetching": false,
  "mockResources": Object {
    "errors": Array [
      "This is an initial error",
      "This is another error",
    ],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;

exports[`test will update the state after the initial action 1`] = `
Object {
  "mockResources": Object {
    "errors": Array [],
    "isFetching": true,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {},
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;

exports[`test will update the state after the success action 1`] = `
Object {
  "mockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": "2016-01-01T00:00:00.000Z",
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {
      "resourceId": Object {
        "id": "resourceId",
        "prop1": "initial prop1",
        "prop2": "initial prop2",
        "relatedResource": "relatedResourceId",
      },
    },
    "resourcesBeingFetched": Array [],
    "selectedResource": "resourceId",
    "selectedResourceList": null,
  },
  "relatedMockResources": Object {
    "errors": Array [],
    "isFetching": false,
    "lastCreatedAt": null,
    "lastDeletedAt": null,
    "lastListedAt": null,
    "lastRetrievedAt": null,
    "lastUpdatedAt": null,
    "resourceLists": Object {},
    "resourceListsBeingFetched": Array [],
    "resources": Object {
      "relatedResourceId": Object {
        "id": "relatedResourceId",
        "relatedProp1": "initial relatedProp1",
      },
    },
    "resourcesBeingFetched": Array [],
    "selectedResource": null,
    "selectedResourceList": null,
  },
}
`;
